bionic/deriver.py

```diff
@@ -460,46 +411,194 @@ def is_resolved(self):
+        assert len(self._ready_entries) != 0
+        return self._ready_entries.pop()
+
+    def _wait_on_in_progress_entries(self):
+        "Waits on any in-progress entry to finish."
+        futures = [
+            entry.future
+            for entry in self._in_progress_entries.values()
+            if entry.future is not None
         ]
-        blockage = TaskBlockage(blocked_state, blocking_tks)
+        finished_futures, _ = wait(futures, return_when=FIRST_COMPLETED)
+        for finished_future in finished_futures:
+            task_key = finished_future.result()
+            entry = self._entries_by_task_key[task_key]
+            entry.state.sync_after_subprocess_completion()
+            self._mark_entry_completed(entry)
+
+    def _mark_entry_ready(self, ready_entry):
+        assert (
+            ready_entry.stage == EntryStage.READY
+            or ready_entry.stage == EntryStage.BLOCKED
+        )
+
+        ready_entry.stage = EntryStage.READY
+        self._ready_entries.append(ready_entry)
+
+    def _mark_entry_blocked(self, blocked_entry):
+        assert blocked_entry.stage == EntryStage.READY
+
+        blocked_entry.stage = EntryStage.BLOCKED
+        blocking_tks = blocked_entry.state.incomplete_dep_task_keys()
+        blockage = EntryBlockage(blocked_entry, blocking_tks)
         for blocking_tk in blocking_tks:
             self._blockage_lists_by_blocking_tk[blocking_tk].append(blockage)
 
-    def get_unblocked_by(self, completed_state):
-        """
-        Records the fact that a task state is complete, and yields all task states
-        that are newly unblocked.
-        """
+        self._ready_blocking_entries(blocked_entry)
+
+    def _mark_entry_in_progress(self, in_progress_entry, future):
+        assert in_progress_entry.stage == EntryStage.READY
+        assert in_progress_entry.future is None
+        assert in_progress_entry not in self._in_progress_entries
+
+        in_progress_entry.stage = EntryStage.IN_PROGRESS
+        in_progress_entry.future = future
+        self._in_progress_entries[
+            in_progress_entry.state.task.keys[0]
+        ] = in_progress_entry
 
-        for completed_tk in completed_state.task.keys:
+    def _mark_entry_completed(self, completed_entry):
+        assert completed_entry.stage == EntryStage.READY or EntryStage.IN_PROGRESS
+
+        if completed_entry.stage == EntryStage.IN_PROGRESS:
+            completed_entry.future = None
+            del self._in_progress_entries[completed_entry.state.task.keys[0]]
+
+        completed_entry.stage = EntryStage.COMPLETED
+        self._unblock_entries(completed_entry)
+
+    def _ready_blocking_entries(self, blocked_entry):
+        incomplete_dep_states = blocked_entry.state.incomplete_dep_states()
+        for incomplete_dep_state in incomplete_dep_states:
+            incomplete_entry = self._get_or_create_entry_for_state(incomplete_dep_state)
+            if incomplete_entry.stage == EntryStage.READY:
```

Comment:
It seems weird to me to call `_mark_entry_ready` on an entry that's already `READY`. Do you think we could make `entry.stage == READY` synonymous with `entry in self._ready_entries`?

Maybe it would be helpful to introduce a new stage or stages for entries that have just been created (and should be put on the ready stack) or have just been popped off the ready stack (and are about to be processed).

Reply:
This felt a little weird to me too but it's because `READY` is the default stage. I'll add a `NEW` stage. 

Also, now that I read this again, I feel that `ready` term is a bit overloaded in the whole runner. I'll rename `_ready_entries` to `_entries_to_process`.